jep/0000/README.adoc

```diff
@@ -0,0 +1,264 @@
+
+On https://github.com/jenkinsci/jenkins/blob/09f0269e87625491d7d897ba0e878a1f7fa31de4/core/src/main/resources/lib/layout/pageHeader.jelly[`pageHeader.jelly`], we will make it more modular by injecting a (customizable) `headerContent.jelly` that will provide the content of the header.
+
+Thus, `pageHeader.jelly` should look like:
+
+```xml
+<?jelly escape-by-default='true'?>
+<j:jelly xmlns:j="jelly:core" xmlns:st="jelly:stapler" xmlns:d="jelly:define" xmlns:l="/lib/layout" xmlns:i="jelly:fmt" xmlns:x="jelly:xml">
+	<j:invokeStatic var="header" className="jenkins.views.Header" method="get"/>
+	<st:include it="${header}" page="headerContent.jelly"/>
+</j:jelly>
+```
+
+The `get()` method is provided in the extension point `jenkins.views.Header` and it will retrieve the available headers via `Extension` lookup that are enabled and for those obtained it will provide the one with max priority (ordinal value).
+
+In terms of simplicity, proposal aims to use only one extension to do a full replacement of the header.
+
+== Motivation
+
+As mentioned before, Jenkins does not provide a customization mechanism for the header.
+
+The unique existing approach based on the https://plugins.jenkins.io/simple-theme-plugin/[simple-theme-plugin] has limited capabilities (see Reasoning section for more details).
+
+It makes it difficult:
+
+* to provide branding capabilities to include custom logos, styles or other elements, and
+* to customize some functionality or include an additional one to some of the elements, like the search bar.
+
+Sometimes, these limited customization capabilities can be a barrier to Jenkins adoption inside enterprises.
+
+Having the ability to customize the header (not only from the User Interface point of view) will help avoid that situation.
+
+== Reasoning
+
+Jenkins uses `pageHeader.jelly` file to specify the content of the header. Although it is valid, if a user wants to modify the header to perform some branding operation, Jenkins header branding capabilities are limited.
+
+Some plugins, like: https://plugins.jenkins.io/simple-theme-plugin/[simple-theme-plugin], allow Jenkins users to customize some parts like CSS and/or Javascript.
+On the other hand, if a Jenkins user wants to customize/modify some additional business functionality on some menus and search bar, then there is no approach beyond updating/overriding `pageHeader.jelly` from the Jenkins core, which would be a problem on updating the instance due to conflicts.
+In addition, if the user wants to customize these behaviors it would be good to have those features as REST endpoints.
+
+Also, other alternatives have been evaluated as workaround like the https://github.com/stephenc/diffpatch-maven-plugin[diffpatch-maven-plugin] with no satisfactory results. See the _Reasoning_ section for further details.
```

Comment:
Optional - since we're already inside the "Reasoning" section, I don't think we need to refer readers to the Reasoning section for more details.
```suggestion
Also, other alternatives have been evaluated as workaround like the https://github.com/stephenc/diffpatch-maven-plugin[diffpatch-maven-plugin] with no satisfactory results.
```